go/ai/prompt/prompt.go

```diff
@@ -0,0 +1,391 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prompt
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/aymerick/raymond"
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/core"
+	"github.com/firebase/genkit/go/internal/atype"
+	"github.com/firebase/genkit/go/internal/base"
+	"github.com/firebase/genkit/go/internal/registry"
+	"github.com/invopop/jsonschema"
+)
+
+type Prompt struct {
+	Registry *registry.Registry
+	Action   core.Action[any, *ai.ModelRequest, struct{}]
+	// The name of the prompt.
+	Name string
+	// Prompt description
+	Description string
+	Config
+	// The parsed prompt template.
+	Template *raymond.Template
+	// The original prompt template text.
+	TemplateText string
+}
+
+// Config is optional configuration for a [Prompt].
+type Config struct {
+	// The prompt variant.
+	Variant string
+	// The name of the model for which the prompt is input.
+	// If this is non-empty, Model should be nil.
+	ModelName string
+	// The Model to use.
+	// If this is set, ModelName should be an empty string.
+	Model ai.Model
+	// The System prompt. If this is non-empty, SystemFn should be nil.
+	System string
+	// The System prompt function. If this is set, System should be an empty string.
+	SystemFn PromptFn
+	// The User prompt. If this is non-empty, PromptFn should be nil.
+	Prompt string
+	// The User prompt function. If this is set, Prompt should be an empty string.
+	PromptFn PromptFn
+	// The messages to add to the prompt. If this is set, MessagesFn should be an empty.
+	Messages []*ai.Message
+	// The messages function. If this is set, Messages should be an empty.
+	MessagesFn MessagesFn
+	// Override the render function
+	RenderFn RenderFn
+	// The tools to use.
+	Tools []ai.Tool
+	// Details for the model.
+	GenerationConfig *ai.GenerationCommonConfig
+	// Schema for input variables.
+	InputSchema *jsonschema.Schema
+	// Default input variable values
+	DefaultInput map[string]any
+	// Desired output format.
+	OutputFormat ai.OutputFormat
+	// Desired output schema, for JSON output.
+	OutputSchema *jsonschema.Schema
+	// Arbitrary metadata.
+	Metadata map[string]any
+	// ToolChoice is the tool choice to use.
+	ToolChoice ai.ToolChoice
+	// MaxTurns is the maximum number of turns.
+	MaxTurns int
+	// ReturnToolRequests is whether to return tool requests.
+	ReturnToolRequests bool
+}
+
+type PromptFn func(context.Context, any) (string, error)
```

Comment:
```suggestion
type PromptFn = func(context.Context, any) (string, error)
```

Let's use type aliases for all of these unless we need to add methods to the types themselves.